### PR TITLE
feat(storage): add base types for storage browser

### PR DESCRIPTION
**Pull request by AllanZhengYP**
#### Description of changes
Add types so far for StorageBrowser.

#### Description of how you validated changes
No implemenation added.

#### Checklist

- [x] PR description included
- [ ] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/storage/src/storage-browser/types.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	// the S3 data through this application.
+	readonly applicationArn?: string;
+}
+
+/**
+ * @internal
+ */
+export interface ListLocationsOutput<T extends LocationAccess> {
+	locations: T[];
+	nextToken?: string;
```
Comment (calebpollman) [inline packages/storage/src/storage-browser/types.ts:50]: Why not `nextToken: string | undefined`? `?` implies an optional property, but return values are not optional

File: packages/storage/src/storage-browser/types.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	// prefix: s3://<bucket>/<prefix-with-path>*
+	// object: s3://<bucket>/<prefix-with-path>/<object>
+	readonly type: LocationAccessType;
+}
+
+export interface AccessGrant extends LocationAccess {
+	// The Amazon Resource Name (ARN) of an AWS IAM Identity Center application associated with your
+	// Identity Center instance. If the grant includes an application ARN, the grantee can only access
+	// the S3 data through this application.
+	readonly applicationArn?: string;
```
Comment (calebpollman) [inline packages/storage/src/storage-browser/types.ts:42]: Same feedback as https://github.com/aws-amplify/amplify-js/pull/13528/files#r1649234243

File: packages/storage/src/providers/s3/types/options.ts [hunk 3]
```diff
@@ -10,12 +11,34 @@ import {
...
+ * @internal
+ */
+export type Permission = 'READ' | 'READWRITE' | 'WRITE';
+
+/**
+ * @internal
+ */
+export type LocationCredentialsProvider = (options: {
+	forceRefresh?: boolean;
+	locations: { bucket: string; path: string }[];
```
Comment (calebpollman) [inline packages/storage/src/providers/s3/types/options.ts:24]: `LocationCredentialsProvider` implies a single location right?

File: packages/storage/src/storage-browser/types.ts (hunk 2 above)
Comment (AllanZhengYP) [reply packages/storage/src/storage-browser/types.ts:42]: Service team confirmed this is indeed optional. Only Access Grants through Identity Center will have this field. AG through IAM will not have it.

File: packages/storage/src/storage-browser/types.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+	// bucket: s3://<bucket>/*
+	// prefix: s3://<bucket>/<prefix-with-path>*
+	// object: s3://<bucket>/<prefix-with-path>/<object>
+	readonly type: LocationAccessType;
+}
+
+export interface AccessGrant extends LocationAccess {
+	// The Amazon Resource Name (ARN) of an AWS IAM Identity Center application associated with your
+	// Identity Center instance. If the grant includes an application ARN, the grantee can only access
+	// the S3 data through this application.
```
Comment (israx) [inline packages/storage/src/storage-browser/types.ts:41]: Should we change this to a TS doc format instead?

File: packages/storage/src/storage-browser/types.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+	readonly applicationArn: string | undefined;
+}
+
+/**
+ * @internal
+ */
+export interface ListLocationsOutput<T extends LocationAccess> {
+	locations: T[];
+	nextToken: string | undefined;
```
Comment (israx) [inline packages/storage/src/storage-browser/types.ts:50]: This output is not consistent with the output of the `list` API. Is there any reason to not make it optional ?

File: packages/storage/src/storage-browser/types.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+export type LocationCredentialsHandler = (
+	input: LocationAccess,
+) => Promise<{ credentials: AWSCredentials; scope?: string }>;
+
+export interface LocationCredentialsStore {
+	// Get location-specific credentials. It uses a cache internally to optimize performance when
+	// getting credentials for the same location. It will refresh credentials if they expire or
+	// when forced to.
+	// If specific credentials scope `forLocation` is omitted, the store will attempt to resolve
+	// locations-specific credentials from the input bucket and full path.
```
Comment (israx) [inline packages/storage/src/storage-browser/types.ts:66]: ditto about changing it to the TS doc format

File: packages/storage/src/storage-browser/types.ts [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+	getProvider(forLocation?: {
+		scope: string;
+		permission: Permission;
+	}): LocationCredentialsProvider;
+	// Invalidate cached credentials and force subsequent calls to get location-specific
+	// credentials to throw. It also makes subsequent calls to `getCredentialsProviderForLocation`
+	// to throw.
```
Comment (israx) [inline packages/storage/src/storage-browser/types.ts:73]: ditto about changing it to the TS doc format

**Review by israx — APPROVED** [review]

**Review by calebpollman — APPROVED** [review]

File: packages/storage/src/storage-browser/listCallerAccessGrants.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+import { AccessGrant, CredentialsProvider, ListLocationsOutput } from './types';
+
+export interface ListCallerAccessGrantsInput {
+	accountId: string;
+	credentialsProvider: CredentialsProvider;
+	region: string;
+	options?: {
+		nextToken?: string;
+		// Default to 100; If > 1000, API will make multiple API calls.
```
[the remark this answers is not in the document]
Comment (AllanZhengYP) [reply packages/storage/src/storage-browser/listCallerAccessGrants.ts:12]: We don't plan to surface this to users. Only the adapter that will surfaced to the UI team.

**Pull request merged**